internal/internal_workflow_execution_options.go

```diff
@@ -0,0 +1,149 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"fmt"
+
+	deploymentpb "go.temporal.io/api/deployment/v1"
+	workflowpb "go.temporal.io/api/workflow/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+)
+
+type (
+	// UpdateWorkflowExecutionOptionsRequest is a request for Client.UpdateWorkflowExecutionOptions.
+	// NOTE: Experimental
+	UpdateWorkflowExecutionOptionsRequest struct {
+		// ID of the workflow.
+		WorkflowId string
+		// Running execution for a workflow ID. If empty string then it will pick the last running execution.
+		RunId string
+		// WorkflowExecutionOptions specifies options for a target workflow execution. Fields not in
+		// UpdatedFields are ignored.
+		WorkflowExecutionOptions WorkflowExecutionOptions
+		// Field names in WorkflowExecutionOptions that will be updated.
+		// When it includes a field name, but the corresponding WorkflowExecutionOptions field has not been set,
+		// it will remove previous overrides for that field.
+		// It panics when it includes a field name not in WorkflowExecutionOptions.
+		// An empty UpdatedFields never modifies WorkflowExecutionOptions.
+		UpdatedFields []string
```

Comment:
Is there a reason we chose this approach to exclude fields over say making the members of `WorkflowExecutionOptions` pointers and treating `nil` as unset

Reply:
I'm following closely the proto api that uses a FieldMask. In general, I've been trying to avoid pointers in the api, and be more explicit about what to change, it is sometimes easy to forget that nil means delete otherwise...

Reply:
My suggestion was a `nil` pointer would leave a field unchanged , not clear it.

Reply:
How do you clear things? Do you use a tombstone?

Reply:
Well here you say 

> // When it includes a field name, but the corresponding WorkflowExecutionOptions field has not been set,
		// it will remove previous overrides for that field.
		
So you are OK relying on the default value to mean clear the field , you can do the same if it takes a pointer.

Reply:
Sure, but I have the other safety net of listing keys that I will delete. I can see people reusing a data structure with the pointers, and making mistakes because they did not clear all of them... @cretz What do you think about a separate selector array vs pointers?

Reply:
We need to decide now how we're going to to do partial updates because `UpdateActivityOptions` (or `UpdateActivityOptionsById`) is coming with the same problem. I would do partial updates in the way the language handles it best. In Go unfortunately there is not one way, but this `UpdatedFields` is probably the least common and IMO we should not expose it.

I would either recommend using pointers or making an update options builder. Pointers works fine.

> I can see people reusing a data structure with the pointers, and making mistakes because they did not clear all of them

We should not use existing data structures in a partial update model because of this concern. So every data type for updating should not reuse the non-updating form on partial update things (this is not a full PUT-with-conflict-token type of API). So either inline the options update fields here or make the data type update-options specific.

Reply:
I'm rewriting it with pointers. We only allow "top level object" partial updates, not arbitrary zipping, so I think that as long as the top level container is a different type we should be ok blocking reuse...